KAFKA-15663, KAFKA-15794: Telemetry reporter and request handling (KIP-714)

apoorvmittal10 opened the pull request:
Implements ClientTelemetryReporter which manages the lifecycle for client metrics collection. The reporter also defines  TelemetrySender which will be used by Network clients to send API calls to broker.

Details about the handling of request and response are covered in KIP-714. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-15663, KAFKA-15794: telemetry reporter and request handling

apoorvmittal10 commented:
Build for the PR depends on PR: https://github.com/apache/kafka/pull/14620

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryUtils.java:
```diff
@@ -0,0 +1,202 @@
+        return notEmptyString(metadata, MetricsContext.NAMESPACE);
+    }
+
+    public static boolean notEmptyString(Map<String, ?> m, String key) {
```
Can we rename this for clarity? It seems to be checking the validity of the config given a key.

apoorvmittal10 replied:
Done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,922 @@
+         */
+        Objects.requireNonNull(this.rawOriginalConfig, "configure() was not called before contextChange()");
+        if (!metricsContext.contextLabels().containsKey(MetricsContext.NAMESPACE)) {
+            log.error("_namespace not found in metrics context. Metrics collection is disabled");
```
does this need to be an error?

apoorvmittal10 replied:
Moved to warn, done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,922 @@
+        collectors.forEach(MetricsCollector::stop);
+
+        if (!metricsProvider.validate(metricsContext, rawOriginalConfig)) {
+            log.error("Validation failed for {} context {}, skip starting collectors", metricsProvider.getClass(), metricsContext.contextLabels());
```
similarly, can we just warn the user?

apoorvmittal10 replied:
Moved to warn, done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,922 @@
+
+    @Override
+    public void close() {
+        log.info("Stopping ClientTelemetryReporter");
```
trace

apoorvmittal10 replied:
Moved to debug, done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,922 @@
+
+        @Override
+        public void close() {
+            log.info("close telemetry sender for client telemetry reporter instance");
```
trace

apoorvmittal10 replied:
Moved to debug, done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,922 @@
+        }
+
+        private void initiateClose(Duration timeout) {
+            log.info("initiate close for client telemetry, check if terminal push required");
```
trace

apoorvmittal10 replied:
Moved to debug, done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,922 @@
+            log.info("initiate close for client telemetry, check if terminal push required");
+
+            long timeoutMs = timeout.toMillis();
+            if (timeoutMs < 0) {
```
I think Duration cannot be < 0 : 
```
The range of a duration requires the storage of a number larger than a long. To achieve this, the class stores a long representing seconds and an int representing nanosecond-of-second, which will always be between 0 and 999,999,999
```

apoorvmittal10 replied:
Removed the check, make sense. Done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryUtils.java:
```diff
@@ -0,0 +1,202 @@
+        int pushIntervalMs;
+        String reason;
+
+        switch (Errors.forCode(errorCode)) {
```
nit about clean up:

 error = Errors.forCode(errorCode)
switch(error) {
  reason = error.message();
  break;
}

apoorvmittal10 replied:
Make sense, done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryUtils.java:
```diff
@@ -0,0 +1,202 @@
+            }
+            default: {
+                reason = "Unwrapped error code";
+                log.error("Error code: {}, reason: {}. Unmapped error for telemetry, disable telemetry.",
```
Not sure if we need a reason here.  Can we just say Error code: {}. Unexpected error ...

philipnee replied:
Also, is the client going to retry on the unexpected case? Or should it throw back to the user.

philipnee replied:
The next error message: log.debug("Error code: {}, reason: {}. Retry automatically in {} ms.", errorCode, reason, pushIntervalMs);

But we aren't actually trying to retry no?

apoorvmittal10 replied:
Thanks for pointing out, I have tried to improve this. Please check.

philipnee commented:
Hi @apoorvmittal10 - I left some questions while making the first pass of the PR. They are all aesthetics.  Will follow up with more questions. Thanks!

apoorvmittal10 committed:
Updated code and addressed review comments

apoorvmittal10 commented:
> Hi @apoorvmittal10 - I left some questions while making the first pass of the PR. They are all aesthetics. Will follow up with more questions. Thanks!

Thanks @philipnee , I have addressed the comments. Can you please re-review.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+                case PUSH_NEEDED: {
+                    String apiName = (localState == ClientTelemetryState.SUBSCRIPTION_NEEDED) ? ApiKeys.GET_TELEMETRY_SUBSCRIPTIONS.name : ApiKeys.PUSH_TELEMETRY.name;
+                    long timeRemainingBeforeRequest = localLastRequestMs + localIntervalMs - now;
+                    if (timeRemainingBeforeRequest < 0) {
```
A tiny nit, but I would say that if it's <= 0, then the time has elapsed. In practice, changing the condition to <= only affects the debug log message.

apoorvmittal10 replied:
thanks, done.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+            // If the error code indicates that the interval ms needs to be updated as per the error
+            // code then update the interval ms and state so that the subscription can be retried.
+            if (errorIntervalMsOpt.isPresent()) {
+                // Update the state from subscription in progress to subscription needed as the error
```
SUBSCRIPTION_IN_PROGRESS and SUBSCRIPTION_NEEDED would be clearer to read matching the code.

apoorvmittal10 replied:
Done.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+                    // If we have a non-negative timeout and no-subscription, let's wait for one to be retrieved.
+                    log.debug("Waiting for telemetry subscription containing the client instance ID with timeoutMillis = {} ms.", timeoutMs);
+                    try {
+                        if (!subscriptionLoaded.await(timeoutMs, TimeUnit.MILLISECONDS)) {
```
Isn't the await here potentially a deadlock?

apoorvmittal10 replied:
I may be missing something but doesn't seems like, I have added a test case as well to validate if we can get the clientInstanceId when concurrent calls are made.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+            int localIntervalMs;
+
+            try {
+                lock.readLock().lock();
```
A more usual pattern for using this kind of lock with try-finally is to put the lock operation just before the try.

apoorvmittal10 replied:
Thanks for pointing out, done.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+        }
+
+        private Optional<Builder<?>> createPushRequest(ClientTelemetrySubscription localSubscription) {
+            if (localSubscription == null) {
```
I would put this if block inside the lock because it does actually attempt a state change.

apoorvmittal10 replied:
The state change is protected by writelock hence I avoided same here. Please let me know if I am missing something here.

AndrewJSchofield replied:
This is true. It is safe. In terms of style, I prefer the entire method to be inside the write lock but happy for you to leave it as it is.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
Thanks for review @AndrewJSchofield, @philipnee. I have addressed the feedback.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryEmitter.java:
```diff
@@ -0,0 +1,64 @@
+
+    public ClientTelemetryEmitter(Predicate<? super MetricKeyable> selector, boolean deltaMetrics) {
+        this.selector = selector;
+        this.emitted = new ArrayList<>();
```
Was just going back to `MetricsEmitter` and JavaDoc says:
```
An {@code MetricsEmitter} is stateless 
```

But `emitted` is clearly used to accumulate state. Can you elaborate of this difference?

Given the implementation, I am actually not sure what "emitted" means here -- does it not mean sent to the broker? Does it instead mean, metric was "emitted' by the client code and thus this class actually _collects/buffers_ metrics until the next interval to them them to the brokers passed? If this is correct, is `MetricEmitter` a good name?

apoorvmittal10 replied:
Thanks for pointing put @mjsax. I earlier thought to make this class stateless but realized it actually depends on the usage. Currently the way we export metrics from the client is dependent on NetworkClient as that manages the data transmission to broker. But we can have potentially another implementation of MetricsEmitter to emit metrics directly to some external system i.e. say directly to any local opentelemetry client.

I have updated the docs comment on `MetricsEmitter`.

Please let me know your thoughts.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -231,7 +231,7 @@
 
     private final Logger log;
-    private static final String JMX_PREFIX = "kafka.producer";
+    public static final String JMX_PREFIX = "kafka.producer";
```
This is a public API change, right? Is this covered in the KIP? Do we want to make it public (we could also move this to an internal class).

apoorvmittal10 replied:
I have reverted this changed.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryUtils.java:
```diff
@@ -0,0 +1,202 @@
+        Errors error = Errors.forCode(errorCode);
+        switch (error) {
+            case INVALID_REQUEST:
+            case INVALID_RECORD: {
```
Is it common in client code base to use `{ }` for case blocks? -- Never say this before in Kafka code base. Just a question of code style.

apoorvmittal10 replied:
You are right this is not common in clients code, I removed the blocks.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,134 @@
+
+    /**
+     * Validate that all the data required for generating correct metrics is present. The provider
+     * will be disabled if validation fails.
```
> The provider will be disabled if validation fails.

It seems this method does not set any internal flag to disable itself?

apoorvmittal10 replied:
Corrected the comment, rather the method calls it disables the telemetry.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,134 @@
+     */
+    public boolean validate(MetricsContext metricsContext, Map<String, ?> config) {
+        // metric collection will be disabled for clients without a client id (e.g. transient admin clients)
+        return ClientTelemetryUtils.validateResourceLabel(config, CommonClientConfigs.CLIENT_ID_CONFIG) &&
```
Not sure if I can follow? Seems, we require that `client.id` is set on the config -- but why? `client.id` is an optional config and we should be able to report telemetry even if not set? Or do I misunderstand something?

apoorvmittal10 replied:
Though the property `client.id` is optional but ProducerConfig and ConsumerConfig generates a client.id if not defined, which is not the case in AdminClient. Hence the admin client created from CLI (i.e. KafkaConfigs.sh, etc.) should not report any metrics.

apoorvmittal10 replied:
I think you are right as `client.id` will anywyas be present. I have removed the check for `client.id`.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,134 @@
+    public boolean validate(MetricsContext metricsContext, Map<String, ?> config) {
+        // metric collection will be disabled for clients without a client id (e.g. transient admin clients)
+        return ClientTelemetryUtils.validateResourceLabel(config, CommonClientConfigs.CLIENT_ID_CONFIG) &&
+            ClientTelemetryUtils.validateRequiredResourceLabels(metricsContext.contextLabels());
```
I cannot find anything about `_namespace` in the KIP. Can you elaborate?

apoorvmittal10 replied:
`_namespace` is the default metrics context which differentiates between various Kafka Components. I am not sure if we need to call this out in KIP-714 as well. We are re-using the metrics and context exposed for JMX hence `_namespace` is the only differentiator.

https://cwiki.apache.org/confluence/display/KAFKA/KIP-606%3A+Add+Metadata+Context+to+MetricsReporter

cc: @AndrewJSchofield

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,134 @@
+    public static final String CLIENT_RACK = "client_rack";
+    public static final String GROUP_ID = "group_id";
+    public static final String GROUP_INSTANCE_ID = "group_instance_id";
+    public static final String TRANSACTIONAL_ID = "transactional_id";
```
The KIP also list `group.member.id` -- why is it not listed here?

apoorvmittal10 replied:
The KIP says `group member id (if any, consumer)`. I considered `if any` as not required label as that will not be present as standard consumer property rather I think broker generates same. Also I do not find `group member id` is exposed as label for consumer metrics. @AndrewJSchofield Please let me know if my understanding is incorrect or we can do better here.

apoorvmittal10 replied:
I found the way to expose `group_member_id`, I am adding partial changes in this PR to define the label and methods to update labels. But would add the final change in next PR where changes are more aligned with AbstractCoordinator and use of ClientTelemetryReporter.

mjsax replied:
The `if any` is mentioned on the KIP for others, too, that's why I found it confusing. -- Thanks for adding.

I ams not sure so, why more changes need to go into a follow up PR? Can you elaborate?

apoorvmittal10 replied:
`group_member_id` is not exposed as config rather comes as response from join group request to broker. Hence the changes are more aligned with next PR (https://github.com/apache/kafka/pull/14843) where changes are aligned more structurally for Produce, Consumer and AbstractCoordinator.

mjsax replied:
Ah. Thanks! -- I mixed it up with the `group.instance.id` config... My bad.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        // Add client rack label.
+        if (config.containsKey(CommonClientConfigs.CLIENT_RACK_CONFIG)) {
```
The KIP says "if configured" so this makes sense. I am just wondering why we make other mandatory? Would it not imply that metrics cannot be reported if optional config like `group.id`, `group.instance.id`, or `group.member.id` are not set?

apoorvmittal10 replied:
I think the KIP should say if configured for all optional configurations used as labels.
cc: @AndrewJSchofield.

AndrewJSchofield replied:
Yes, @apoorvmittal10 is reading this correctly. These standard labels will be included if set in the config.

mjsax replied:
Ah. I think I read the code incorrectly. Thanks. All good.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,134 @@
+     * @return A fully formed {@link Resource} with all the tags.
+     */
+    public Resource resource() {
+        return this.resource;
```
nit: remove unnecessary `this.`

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+    private static final Logger log = LoggerFactory.getLogger(ClientTelemetryReporter.class);
+
+    public static final int DEFAULT_PUSH_INTERVAL_MS = 5 * 60 * 1000;
+    public static final long MAX_TERMINAL_PUSH_WAIT_MS = 1000;
```
Cannot remember this from the KIP? Can you elaborate? Should it be part of the KIP (just honestly asking; not requesting to add it).

apoorvmittal10 replied:
I have actually changed the way connections should be closed and the timeout should come from front APIs i.e. KafkaProducer, KafkaConsumer, etc.

apoorvmittal10 replied:
Removed this config.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+          If validation succeeds: initialize the provider, start the metric collection task,
+          set metrics labels for services/libraries that expose metrics.
+         */
+        Objects.requireNonNull(this.rawOriginalConfig, "configure() was not called before contextChange()");
```
Should we add this check additionally to `configure(...)` above? -- In the end, the error message is not 100% correct, because `contextChange()` might have been called but with `null` ?

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+        kafkaMetricsCollector = new KafkaMetricsCollector(
+            TelemetryMetricNamingConvention.getClientTelemetryMetricNamingStrategy(
+                telemetryProvider.domain()));
+        collectors.add(kafkaMetricsCollector);
```
Not sure if I fully understand the control flow? Would `collectors` be a list with only a single collector? Or when would a collector be removed?

apoorvmittal10 replied:
As of now the ClientTelemetryReporter has single collector but I have kept the list so this can be extended i.e. if required to collect HostProcess metrics, metrics other than JMX, etc. Please let me know if you think that we shouldn't keep the list rather work just with KafkaMetricsCollector for now.

mjsax replied:
If we don't plan to support mort than one collector via KIP-714, I might be best to remove the list and keep the code simple. We can change it when support for a second collector is added. Otherwise the code is unnecessary complex (and we might never add support for a second collector).

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+
+        @Override
+        public long timeToNextUpdate(long requestTimeoutMs) {
+            final long now = System.currentTimeMillis();
```
We should use a `Time` object and never call `System.currrentTimeMillis()` directly.

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+            final long now = System.currentTimeMillis();
+            final ClientTelemetryState localState;
+            final long localLastRequestMs;
+            int localIntervalMs;
```
The are the other two `final` but this one is not?

apoorvmittal10 replied:
Missed that, done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+                     submission, so wait that amount of the time PLUS the requestTimeout that
+                     is provided.
+                    */
+                    String apiName = (localState == ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS) ? ApiKeys.GET_TELEMETRY_SUBSCRIPTIONS.name : ApiKeys.PUSH_TELEMETRY.name;
```
Do we need to distinguish between regular and final push? If the final push is in progress, should this method actually get called? And if yes, should it in doubt set `timeMs` to max-value to "disable" sending more requests?

apoorvmittal10 replied:
You are right, done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+                    break;
+                }
+                default:
+                    // Should never get to here
```
If it would indicate a bug, should we rather throw an `InvalidStateException()` to surface the bug instead of masking it?

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+            if (localState == ClientTelemetryState.SUBSCRIPTION_NEEDED) {
+                return createSubscriptionRequest(localSubscription);
+            } else if (localState == ClientTelemetryState.PUSH_NEEDED || localState == ClientTelemetryState.TERMINATING_PUSH_NEEDED) {
+                return createPushRequest(localSubscription);
```
IIRC from the KIP, the final push has "terminating" flag set to true? Don't we need to distinguish both cases here?

apoorvmittal10 replied:
This is done later when actually request is constructed.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+
+        @Override
+        public void handleResponse(GetTelemetrySubscriptionsResponse response) {
+            final long now = System.currentTimeMillis();
```
As above: use `Time`

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,914 @@
+
+        @Override
+        public void handleResponse(PushTelemetryResponse response) {
+            final long now = System.currentTimeMillis();
```
use `Time`

apoorvmittal10 replied:
Done.

mjsax reviewed (commented):
Made a first pass.

apoorvmittal10 committed:
Addressing review comments - reporter

apoorvmittal10 committed:
Removing unused imports - reporter

apoorvmittal10 committed:
Removing unneccessary check and added log line - reporter

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,141 @@
+        PRODUCER_CONFIG_MAPPING.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, ClientTelemetryProvider.TRANSACTIONAL_ID);
+
+        CONSUMER_CONFIG_MAPPING.put(ConsumerConfig.GROUP_ID_CONFIG, ClientTelemetryProvider.GROUP_ID);
+        CONSUMER_CONFIG_MAPPING.put(ConsumerConfig.GROUP_INSTANCE_ID_CONFIG, ClientTelemetryProvider.GROUP_INSTANCE_ID);
```
It seem we should add `group.member.id` there, too?

apoorvmittal10 replied:
`group.member.id` is not required to be added in this static mapping, as the properly is not defined in config rather arrives in response from join request. Once the response arrives the `updateMetricsLabel` is called to update resource metrics labels.

mjsax replied:
Might be worth to add a comment?

apoorvmittal10 replied:
Done.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -550,6 +576,41 @@ public void close() {
+            try {
+                // If we never fetched a subscription, we can't really push anything.
+                if (lastRequestMs == 0) {
+                    log.info("Telemetry subscription not loaded, not attempting terminating push");
```
This is `info` while others below are `debug`. Why?

apoorvmittal10 replied:
Missed it, my bad, earlier kept everything at INFO for local testing. Moved to debug.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,937 @@
+        }
+    }
+
+    public synchronized void updateMetricsLabels(Map<String, String> labels) {
```
I cannot find where this is called, and thus it's not clear to me why it's added (similar for `telemetryProvider.updateLabels(...)` which is only called from here as it seems)

apoorvmittal10 replied:
The usage is in the PR: https://github.com/apache/kafka/pull/14843/files to update group_member_id metric label.

mjsax reviewed (commented):
Few minor follow ups.

apoorvmittal10 committed:
Removing collectors list - reporter

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

apoorvmittal10 committed:
Correcting test

apoorvmittal10 committed:
Correcting log level

apoorvmittal10 commented:
Hi @mjsax, thanks a lot for the review. I have addressed the feedback, please let me know if I need to further address anything.

I have updated the PR with other merges and will wait for build for successful completion now.

apoorvmittal10 committed:
Moving client telemetry utils to internal - reporter

apoorvmittal10 commented:
@mjsax I have also moved ClientTelemetryUtils to `internals` package in this PR as well, the class is new and have changes in both PRs.

apoorvmittal10 committed:
Correcting import order - reporter

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

apoorvmittal10 committed:
Added comment

apoorvmittal10 commented:
@mjsax I have added the comment as suggested.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,144 @@
+        resource = resourceBuilder.build();
+    }
+
+    public synchronized void updateLabels(Map<String, String> labels) {
```
i wonder if we should be more consistent with documenting all public methods.

mjsax replied:
Is this actually, public? I am just realizing that the whole class is not mention on the KIP. I think it should go into some `internal` package to begin with?

On the KIP, I can only see `ClientTelemetry`, `ClientTelemetryReceiver`,`ClientTelemetryPayload`, and  being newly added public classes?

But it seems client model is not very strict historically about where to add classes?

apoorvmittal10 replied:
The changes were more specific to `clients` hence kept them in the respective package. I can move them to `org.apache.kafka.common.telemetry.internals`, these are the internal classes.

apoorvmittal10 replied:
I have added the doc, changed the modifier to `package-private`, moved class to internals.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+ * The client telemetry reporter is configured with a {@link ClientTelemetrySender} which is
+ * responsible for sending the client telemetry data to the broker. The client telemetry reporter
+ * will attempt to fetch the telemetry subscription information from the broker and send the
+ * telemetry data to the broker based on the subscription information.
```
could you be more specific about what you meant by "based on the subscription information" - Do you mean we only send telemetric data for the subscribed topics/partitions?

apoorvmittal10 replied:
I have added `i.e. push interval, temporality, compression type, etc.` in comment to make to more specific.

philipnee replied:
great thanks

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+ * <p>
+ *
+ * The full life-cycle of the metric collection process is defined by a state machine in
+ * {@link ClientTelemetryState}. Each state is associated with a different set of operations.
```
Remind me, did you documented the transition in the TelemtryState statement? I think you did, but I want to double check.

apoorvmittal10 replied:
Not yet, planning to do after 3.7 code freeze. I think you mean the way KIP-848 state transition is publically documented? I have created a jira so can track: https://issues.apache.org/jira/browse/KAFKA-15952

philipnee replied:
sounds good - what i meant is since you mentioned the state transition, I wonder if we should just document the transition in the code.

apoorvmittal10 replied:
Yeah, the code has a write up in ClientTelemetryState.java where next states are mentioned.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+ * <p>
+ *
+ * In an unlikely scenario, if a bad state transition is detected, an
+ * {@link IllegalStateException} will be thrown.
```
how do we hanle the exception?

apoorvmittal10 replied:
The error gets currently logged but other client operations do not fail. The log will help determine if there is any incorrect transition so that can be fixed.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+
+    @Override
+    public synchronized void contextChange(MetricsContext metricsContext) {
+        /*
```
nit-picking: below you were using `//` here you are using `/* */` and it doesn't seem to be a huge body of text.  Maybe let's use `//` all together to be more consistent, stylistically.

apoorvmittal10 replied:
Prefer block comments generally for multiple line. I have moved to block comments for the file when comments are in multiple lines.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+    private final Time time;
+
+    private Map<String, Object> rawOriginalConfig;
+    private KafkaMetricsCollector kafkaMetricsCollector;
```
can the collector be optional?

apoorvmittal10 replied:
I think it's better to keep it as like this, wrapping and unwrapping Collector in this case would be extra when Collector can only be null in some error scenarios.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+
+        private Optional<Builder<?>> createPushRequest(ClientTelemetrySubscription localSubscription) {
+            if (localSubscription == null) {
+                log.warn("Telemetry state is {} but subscription is null; not sending telemetry", state);
```
maybe move this below the maybeSetState.  The order makes a bit more sense for me: I was thinking the logical order would be:

1. Unable to transition....
2. [therefore]... not sending telemetry

As you are logging at error level - it seems like we aren't doing anything after a failed state transition. Should we reset the state or something? Would it cause repeated error messages being logged?  If this is not a fatal condition, maybe we could just warn?

apoorvmittal10 replied:
Thanks @philipnee. I think the order is correct as I do want to log the state prior re-setting the state from `mayBeSetState`. 

I agree that the log level shouldn't be `error` and I changed to `warn` if state cannot be updated. If state transition is not successful, which should not be the case ideally, then the error log might appear after push interval is elapsed.

I did think about re-setting to some initial `state` in case of error but if the client state transition is buggy then the problem will re-occur. I have added line to disable telemetry in case of error state transition.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+         * Updates the {@link ClientTelemetrySubscription}, {@link #intervalMs}, and
+         * {@link #lastRequestMs}.
+         * <p>
+         * The entire contents of the method are guarded by the {@link #lock}.
```
not sure this is needed as it is quite apparent 😅

apoorvmittal10 replied:
Make sense, removed.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+                this.subscription = Objects.requireNonNull(subscription);
+                /*
+                 If the subscription is updated for the client, we want to attempt to spread out the push
+                 requests between 50% and 150% of the push interval value from the broker. This helps us
```
this is what the computeStaggeredIntervalMs I assume

philipnee replied:
I think we can add this comment directly above the method

apoorvmittal10 replied:
I find it more appropriate to add the comment here where we are checking if the compute of push interval is required or not. The method comment will mostly talk about what the behaviour of method should be but this comment has additional context. Do let me know if you differ.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,952 @@
+            try (MetricsEmitter emitter = new ClientTelemetryEmitter(localSubscription.selector(), localSubscription.deltaTemporality())) {
+                emitter.init();
+                kafkaMetricsCollector.collect(emitter);
+                payload = createPayload(emitter.emittedMetrics());
```
If you can't create a payload due to some exception, do you return an empty Optional? If so can we move the catch{ } in createPayload here?

apoorvmittal10 replied:
Make sense, done.

philipnee commented:
Hi @apoorvmittal10 thanks for writing this PR - also sorry for the delay but I made the first pass and left some stylistic comments.

apoorvmittal10 committed:
Moved classes to internal

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
> Hi @apoorvmittal10 thanks for writing this PR - also sorry for the delay but I made the first pass and left some stylistic comments.

Hi @philipnee Thanks for the review. I have addressed the comments.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

philipnee commented on clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryProvider.java:
```diff
@@ -0,0 +1,149 @@
+        if (PRODUCER_NAMESPACE.equals(namespace)) {
+            // Add producer resource labels.
+            PRODUCER_CONFIG_MAPPING.forEach((configKey, telemetryKey) -> {
+                if (config.containsKey(configKey)) {
```
do we need to log debug if the configKey does not exist? This might be useful to debug purpose.

apoorvmittal10 replied:
I think not required as configs are optional, I never needed this log while testing and debugging the complete feature. I ll skip this.

philipnee commented on clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,976 @@
+
+        @Override
+        public long timeToNextUpdate(long requestTimeoutMs) {
+            final long now = time.milliseconds();
```
with unit is better - nowMs

apoorvmittal10 replied:
Though mostly clients code uses `now` for `time.milliseconds()` hence followed the same, there are some occurrences on `nowMs` as well. I do not have any strong opinion on either hence changed it to `nowMs`.

philipnee commented on clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,976 @@
+                     is provided.
+                    */
+                    apiName = (localState == ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS) ? ApiKeys.GET_TELEMETRY_SUBSCRIPTIONS.name : ApiKeys.PUSH_TELEMETRY.name;
+                    timeMs = requestTimeoutMs;
```
Need a bit of clarification here, it was stated:
```
If another telemetry API is in-flight, then {@code timeoutMs} should be returned as the
     * maximum wait time.
```

I assume maximum wait time is more than requestTimeoutMs?

apoorvmittal10 replied:
If the telemetry request is in flight then method returns the `requestTimeoutMs` provided from `NetworkClient`, after which the next request can be retried. This time is the maximum wait time for the current call to `timeToNextUpdate`.

philipnee commented on clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,976 @@
+                    throw new IllegalStateException("Unknown telemetry state: " + localState);
+            }
+
+            log.debug("For telemetry state {}, returning the value {} ms; {}", localState, timeMs, msg);
```
would it be more descriptive to say: returning the value -> time for the next push in {} ms ?

apoorvmittal10 replied:
The last param in log is the `msg` which is computed in switch statement and I think it's quite descriptive. The `msg` actually covers your suggestion.

philipnee commented on clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java:
```diff
@@ -0,0 +1,976 @@
+            }
+        }
+
+        private int computeStaggeredIntervalMs(int intervalMs) {
```
it's a bit personal, but when it comes to testing random value, it would be nice to test it with jitter = 0 because then we can expect the same result.  This is mainly for testing the nextTimeToPushMs() function.

apoorvmittal10 replied:
Though it's not exponential backoff jitter where if jitter is 0 then the same result will be returned, rather it's a jitter for upper and lower bound where value of the staggered time should be. I have changed the method to take lower and upper bound params and added a direct test as well, earlier was testing this method only from `timeToNextUpdate`

philipnee commented on clients/src/test/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporterTest.java:
```diff
@@ -0,0 +1,573 @@
+
+        assertTrue(telemetrySender.maybeSetState(ClientTelemetryState.PUSH_NEEDED));
+        long time = telemetrySender.timeToNextUpdate(100);
+        assertTrue(time > 0 && time >= 0.5 * time && time <= 1.5 * time);
```
i left a comment about testing with deterministic value above.

apoorvmittal10 replied:
Replied above, resolving comment.

philipnee commented on clients/src/test/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporterTest.java:
```diff
@@ -0,0 +1,573 @@
+        assertTrue(requestOptional.get().build() instanceof GetTelemetrySubscriptionsRequest);
+        GetTelemetrySubscriptionsRequest request = (GetTelemetrySubscriptionsRequest) requestOptional.get().build();
+
+        GetTelemetrySubscriptionsRequest expectedResult = new GetTelemetrySubscriptionsRequest.Builder(
```
Hey I just realized - we should file a separate jira to test different request versions.  This helps to verify the correctness of the request. It's really simple, basically you just need to verify the fields are correct.  There's an example test `public void testValidateConsumerGroupHeartbeatRequest(final short version)`

apoorvmittal10 replied:
Filed: https://issues.apache.org/jira/browse/KAFKA-15958

philipnee commented on clients/src/test/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporterTest.java:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    @Test
+    public void testHandleResponsePushTelemetryErrorResponse() {
```
Hey would it be possibe to parametrize the test?

apoorvmittal10 replied:
As the states are different from where the next transition should be hence skipped the parametrized tests.

philipnee commented on clients/src/test/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporterTest.java:
```diff
@@ -0,0 +1,573 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ClientTelemetryReporterTest {
```
should we test the case where the subscription changes when creating a push request?

apoorvmittal10 replied:
I am not sure if I got the comment correctly. If the comment is for adding tests case when push telemetry receives subscription error them that's covered under `testHandleResponsePushTelemetryErrorResponse`. However I have added another test case now which verifies the change in subscription (also change in temporality) when handling get subscriptions as well.

philipnee replied:
thanks that's what i meant.

philipnee commented:
Hi @apoorvmittal10 Thanks for the time to address my comments - I made another pass and added some more comments, mostly nit-picking.  After this I think we are good to go.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
> Hi @apoorvmittal10 Thanks for the time to address my comments - I made another pass and added some more comments, mostly nit-picking. After this I think we are good to go.

Hi @philipnee thanks for the re-review. I have addressed the feedback.

philipnee reviewed (approved):
Thanks @apoorvmittal10 for the PR.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-client

apoorvmittal10 commented:
Another build succeeded with flaky tests.
